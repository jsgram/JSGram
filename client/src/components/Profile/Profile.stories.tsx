import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Provider } from 'react-redux';
import { store } from '../../App';
import { MemoryRouter } from 'react-router-dom';
import Profile from './index';

export const profileState = {
    urlUsername: 'Maks',
    loggedId: '1a',
    loggedUsername: 'Maks',
    loggedUser: {
        isAdmin: false,
        loggedId: '1a',
        loggedPhotoPath: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
        loggedUsername: 'Maks',
    },
    user: {
        description: 'User description',
        email: 'maks@gmail.com',
        followers: [1, 2, 3, 4, 5],
        following: [1, 2, 3],
        fullName: 'Maks',
        photo: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
        posts: 23,
        username: 'Maks',
        _id: '1a',
    },
    loaded: false,
    loading: false,
    loadFollow: false,
};

const profileActions = {
    getUser: action('getUser'),
    followUser: action('followUser'),
    unfollowUser: action('unfollowUser'),
    deletePhoto: action('deletePhoto'),
    resetPosts: action('resetPosts'),
    getPostsAsync: action('getPostsAsync'),
    deleteUser: action('deleteUser'),
    emitNewNotificationSocket: action('emitNewNotificationSocket'),
};

storiesOf('Profile', module)
    .addDecorator((story: any): any => <Provider store={store}>{story()}</Provider>)
    .addDecorator((story: any): any => <MemoryRouter>{story()}</MemoryRouter>)
    .add('profile', () => <Profile {...{...profileState, ...profileActions}}/>);
