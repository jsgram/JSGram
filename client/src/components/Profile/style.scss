@import '../../styles/variables';

.profile {
  &-name {
    font-weight: 700;
    font-size: 1.3rem;
  }

  &-info {
    flex-direction: column;
  }

  .description {
    max-width: 320px;
  }   

  .following-button {
    background-color: $color-white;
    border-style: none;
  }
}

.following-button {
  &:focus {
    outline: 0;
  }
}


 
