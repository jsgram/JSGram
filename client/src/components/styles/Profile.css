.profile {
  margin: 0 auto;
}

@media screen and (max-width: 480px) {
  .profile .btn {
    font-size: 0.75rem;
  }
  .profile .profile-name {
    font-size: 1rem;
  }
  .profile .followers {
    font-size: 0.8rem;
  }
  .profile .followers a {
    font-size: 0.8rem;
  }
}

.profile-name {
  font-weight: 700;
  font-size: 1.3rem;
}

.profile-info {
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
}

.followers a {
  text-decoration: none;
  color: #000;
  font-size: 1rem;
}

.followers a:hover {
  text-decoration: none;
  color: #000;
}
/*# sourceMappingURL=Profile.css.map */