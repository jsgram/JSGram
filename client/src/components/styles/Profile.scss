body {
  background-color: #f4f4f4;
}

.profile {
  margin: 0 auto;
}

.profile-name {
  font-weight: 700;
  font-size: 1.3rem;
}

.button {
  font-size: 0.9rem;
}

.profile-info {
  flex-direction: column;
}

a {
  text-decoration: none;
  color: #000;
  font-size: 1rem;
}

a:hover {
  text-decoration: none;
  color: black;
}

@media screen and (max-width: 383px) {
  .profile-name {
    font-size: 1rem;
  }

  .button {
    font-size: 0.7rem;
  }

  .followers {
    font-size: 0.7rem;
  }

  a {
    font-size: 0.7rem;
  }
}






