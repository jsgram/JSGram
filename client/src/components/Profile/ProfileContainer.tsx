import React from 'react';
import { connect } from 'react-redux';
import Profile from './Profile';
import { getUser } from '../../store/profile/actions';

interface IStateToProps {
    user: any;
    loading: boolean;
}

interface IDispatchToProps {
    getUser: () => void;
}

interface IState {
    profile: {
        user: any;
        loading: boolean;
    };
}

class ProfileContainer extends React.Component<IStateToProps & IDispatchToProps> {
    public render(): any {
        return (
            <div className='container'>
                <Profile
                    user={this.props.user}
                    loading={this.props.loading}
                    getUser={this.props.getUser}
                />
            </div>
        );
    }
}

const mapStateToProps = (state: IState): { user: any, loading: boolean } => {
    return {
        user: state.profile.user,
        loading: state.profile.loading,
    };
};

const mapDispatchToProps = {
    getUser,
};

export default connect(mapStateToProps, mapDispatchToProps)(ProfileContainer);
