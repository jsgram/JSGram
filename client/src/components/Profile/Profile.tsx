import React from 'react';
import '../styles/style.scss';
import { Instagram } from 'react-content-loader';
import { Button } from 'reactstrap';
import '../styles/Profile.scss';

export interface IUserData {
    first_name: string;
    last_name: string;
    avatar: string;
}

interface IFormProps {
    getUser: () => void;
    user: IUserData;
    loaded: boolean;
}

export default class Profile extends React.Component<IFormProps> {

    public state: {loaded: boolean} = {
        loaded: false,
    };
    public timerHandle: any = 0;

    public componentDidMount(): void {
        this.props.getUser();

    }

    public componentDidUpdate(prevProps: any): void {
        if (prevProps.loaded !== this.props.loaded && this.props.loaded) {
            this.timerHandle = setTimeout(() => {
                this.setState({loaded: true});
                this.timerHandle = 0;
            },
                3000,
            );
        }
    }

    public componentWillUnmount(): void {
        clearTimeout(this.timerHandle);
        this.timerHandle = 0;
    }

    public render(): JSX.Element {
        const {user: {first_name, last_name, avatar}}: any = this.props;
        const {loaded}: {loaded: boolean} = this.state;

        if (!loaded) {
            return (<Instagram/>);
        }
        return (
                <div className='profile-header'>
                    <div className='container'>
                        <div className='row'>
                            <div className='profile d-flex mt-5'>
                                <div className='col-md-4 col-sm-4 col-4 mr-md-4 mr-sm-4'>
                                    <img
                                        src={avatar}
                                        className='img-fluid rounded-circle'/>
                                </div>
                                <div>
                                    <p className='profile-name'>
                                        {first_name} {last_name}
                                        <Button className='bg-dark ml-md-2 ml-sm-2 ml-2 btn'>
                                            Edit profile
                                        </Button>
                                    </p>
                                    <div className='d-flex followers'>
                                        <div>
                                            <a href='#' className='mr-md-4 mr-sm-4 mr-3'>Posts</a>
                                            <p>1000</p>
                                        </div>
                                        <div>
                                            <a href='#' className='mr-md-4 mr-sm-4 mr-3'>Followers</a>
                                            <p className='ml-3'>50m</p>
                                        </div>
                                        <div>
                                            <a href='#'>Following</a>
                                            <p className='ml-4'>12</p>
                                        </div>
                                    </div>
                                    <Button className='btn' color='danger'><i
                                        className='fa fa-plus pr-3'/>Add
                                        Post</Button>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
        );
    }
}
