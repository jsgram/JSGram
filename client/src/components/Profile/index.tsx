import React from 'react';
import '../../styles/style.scss';
import { Instagram } from 'react-content-loader';
import { Button, Spinner, Modal, ModalHeader, ModalBody } from 'reactstrap';
import './style.scss';
import { FollowersData} from '../Followers/FollowersData';
import { Followers } from '../Followers';
import { PopUpModal } from '../PopUp';
import noAvatar from '../../assets/noAvatar.svg';
import Menu from '../Menu';
import { Link } from 'react-router-dom';
import PostContainer from '../../containers/PostContainer';

export interface IUserData {
    posts: number;
    followers: number;
    following: number;
    description: string;
    fullName: string;
    username: string;
    photo: string;
    email: string;
    _id: string;
    getPostsAsync: () => void;
}

export interface IProfileProps {
    urlUsername: string;
    loggedUsername: string;
    user: IUserData;
    loaded: boolean;
    loading: boolean;
    getUser: (username: string) => void;
    deletePhoto: () => void;
    resetPosts: () => void;
    getPostsAsync: (username: string) => void;
}

export default class Profile extends React.Component<IProfileProps> {

    public state: { loaded: boolean, followersModal: boolean, followingModal: boolean, avatarModal: boolean } = {
        loaded: false,
        avatarModal: false,
        followersModal: false,
        followingModal: false,
    };
    public timerHandle: any = 0;

    public componentDidMount(): void {
        this.props.getUser(this.props.urlUsername);
    }

    public componentDidUpdate(prevProps: IProfileProps): void {
        if (prevProps.loaded !== this.props.loaded && this.props.loaded) {
            this.timerHandle = setTimeout(() => {
                this.setState({loaded: true});
                this.timerHandle = 0;
            },
                1500,
            );
        }

        if (this.props.urlUsername !== this.props.user.username && this.props.loaded) {
            this.setState({loaded: false});
            this.props.getUser(this.props.urlUsername);
            this.props.getPostsAsync(this.props.urlUsername);
        }
    }

    public componentWillUnmount(): void {
        clearTimeout(this.timerHandle);
        this.timerHandle = 0;
    }

    public avatarToggle = (): void => {
        this.setState({avatarModal: !this.state.avatarModal});
    }

    public followersToggle = (): void => {
        this.setState({ followersModal: !this.state.followersModal });
    }
    public followingToggle = (): void => {
        this.setState({ followingModal: !this.state.followingModal });
    }

    public render(): JSX.Element {
        const {user: {posts, followers, following, fullName, username, description, photo}}: IProfileProps = this.props;
        const {loaded}: { loaded: boolean } = this.state;

        if (!loaded) {
            return (<Instagram/>);
        }
        return (
            <div
                className='row profile d-flex pt-2 justify-content-lg-center
                justify-content-sm-around justify-content-center'>
                <Menu/>
                <div className='mr-lg-5 mr-3'>
                    {this.props.loading ? <Spinner style={{height: 150, width: 150}} type='grow' color='dark'/> : <img
                        src={photo || noAvatar}
                        className='img-fluid float-right mb-2 avatar-img'
                        alt='avatar'
                        height={150}
                        width={150}
                        onClick={this.avatarToggle}
                    />}
                </div>
                <div className='ml-lg-5 d-sm-block d-flex flex-column'>
                    <p className='profile-name align-self-center'>
                        {username}
                        {this.props.urlUsername === this.props.loggedUsername &&
                        <Link to={`/profile/${this.props.urlUsername}/edit`}>
                            <button className='bg-dark ml-sm-5 ml-3 btn text-white'>
                                Edit profile
                            </button>
                        </Link>
                        }
                    </p>
                    <div className='d-flex followers justify-content-between'>
                        <div>
                            <button className='mr-2 following-button'><b>{posts}</b> posts</button>
                        </div>
                        <div>
                            <button onClick={this.followersToggle} className='mr-2 following-button'>
                                <b>{followers}</b> followers</button>
                        </div>
                        <div>
                            <button onClick={this.followingToggle} className='following-button'>
                                <b>{following}</b> following</button>
                        </div>
                    </div>
                    <div className='description mt-4'>
                        <strong>{fullName}</strong>
                        <p>{description}</p>
                    </div>
                    <Link to='/add-post'>
                        {this.props.urlUsername === this.props.loggedUsername &&
                        <Button className='btn' color='danger'><i
                            className='fa fa-plus pr-3'/>
                            Add Post
                        </Button>
                        }
                    </Link>
                    {this.state.avatarModal && <PopUpModal
                        modal={this.state.avatarModal}
                        toggleModal={this.avatarToggle}
                        loading={this.props.loading}
                        deletePhoto={this.props.deletePhoto}
                        photo={photo}
                    />}
                </div>
                <div className='container'>
                    <PostContainer username={this.props.urlUsername}/>
                </div>
                <Followers toggle={this.followersToggle} modal={this.state.followersModal} title='Followers'/>
                <Followers toggle={this.followingToggle} modal={this.state.followingModal} title='Following'/>
            </div>
        );
    }
}
