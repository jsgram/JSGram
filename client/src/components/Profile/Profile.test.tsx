import Profile from './index';
import * as reactstrap from 'reactstrap';

import { shallow } from 'enzyme';
import React from 'react';

describe('Profile component:', () => {
    let renderer;
    let mockGetUser;

    beforeEach(() => {
        reactstrap.Spinner = jest.fn(() => (<div></div>));
        reactstrap.Button = jest.fn(() => (<div></div>));

        mockGetUser = jest.fn(() => 'somevalue');
        const props = {
            user: {},
            getUser: mockGetUser,
            loaded: true,
        };

        renderer = shallow(<Profile {...props} />);
    });

    test('componentDidMount - success', () => {
        renderer.instance().componentDidMount();
        expect(mockGetUser).toHaveReturnedWith('somevalue');
    });

    test('componentDidUpdate - success', () => {
        renderer.instance().componentDidUpdate({ loaded: false });
        expect(renderer.instance().timerHandle).toBe(6);
    });

    test('componentWillUnmount - success', () => {
        renderer.instance().componentWillUnmount();
        expect(renderer.instance().timerHandle).toBe(0);
    });

    test('render - success', () => {
        // TODO to toHaveLength
        expect(renderer.html().length).toBeGreaterThan(1120);
    });
});
