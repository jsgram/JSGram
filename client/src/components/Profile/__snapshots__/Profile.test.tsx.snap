// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component: render - success 1`] = `ShallowWrapper {}`;
